### PR TITLE
Hybridizing NIST ECC groups with the OQS groups.

### PR DESCRIPTION
**Pull request by anhu**

### CODE REVIEW

File: src/tls.c [hunk 1]
```diff
@@ -7150,6 +7151,72 @@ static const char* OQS_ID2name(int id)
...
     return NULL;
 }
+
+typedef struct OqsHybridMapping {
+    int hybrid;
+    int ecc;
+    int oqs;
+} OqsHybridMapping;
+
+static OqsHybridMapping oqs_hybrid_mapping[] = {
```
Comment (dgarske) [inline src/tls.c:7161]: Please make `const`.

File: src/tls.c [hunk 2]
```diff
@@ -7150,6 +7151,72 @@ static const char* OQS_ID2name(int id)
...
+    {.hybrid = WOLFSSL_P384_KYBER90S768,     .ecc = WOLFSSL_ECC_SECP384R1,
+     .oqs = WOLFSSL_KYBER90S768},
+    {.hybrid = WOLFSSL_P521_KYBER90S1024,    .ecc = WOLFSSL_ECC_SECP521R1,
+     .oqs = WOLFSSL_KYBER90S1024},
+    {.hybrid = 0, .ecc = 0, .oqs = 0}
+};
+
+/* This will map an ecc-oqs hybrid group into its ecc group and oqs group.
+ * If it cannot find a mapping then *oqs is set to group. ecc is optional. */
+static void findEccOqs(int *ecc, int *oqs, int group) {
```
Comment (dgarske) [inline src/tls.c:7193]: Function should have explicit start brace.

File: src/tls.c [hunk 3]
```diff
@@ -7159,13 +7226,17 @@ static const char* OQS_ID2name(int id)
...
 static int TLSX_KeyShare_GenOqsKey(WOLFSSL *ssl, KeyShareEntry* kse)
 {
-    int ret = -1;
+    int ret = 0;
     const char* algName = NULL;
     OQS_KEM* kem = NULL;
     byte* pubKey = NULL;
     byte* privKey = NULL;
+    KeyShareEntry *ecc_kse = NULL;
+    int oqs_group;
```
Comment (dgarske) [inline src/tls.c:7235]: Might consider initializing these, since you don't have return code checking on `findEccOqs`.

File: src/tls.c [hunk 4]
```diff
@@ -7178,35 +7249,77 @@ static int TLSX_KeyShare_GenOqsKey(WOLFSSL *ssl, KeyShareEntry* kse)
         return BAD_FUNC_ARG;
     }
 
-    pubKey = (byte*)XMALLOC(kem->length_public_key, ssl->heap,
-                               DYNAMIC_TYPE_PUBLIC_KEY);
-    privKey = (byte*)XMALLOC(kem->length_secret_key, ssl->heap,
-                                DYNAMIC_TYPE_PRIVATE_KEY);
-    if (pubKey == NULL || privKey == NULL) {
-        WOLFSSL_MSG("memory allocation failure");
+    ecc_kse = XMALLOC(sizeof(*ecc_kse), ssl->heap, DYNAMIC_TYPE_TLSX);
```
Comment (dgarske) [inline src/tls.c:7252]: You must cast this to `(KeyShareEntry*)`

File: src/tls.c [hunk 5]
```diff
@@ -7178,35 +7249,77 @@ static int TLSX_KeyShare_GenOqsKey(WOLFSSL *ssl, KeyShareEntry* kse)
...
-    else {
-        WOLFSSL_MSG("liboqs keygen failure");
-        ret = BAD_FUNC_ARG;
+
+    if (ret == 0 && ecc_group != 0) {
+        ecc_kse->group = ecc_group;
+        ret = TLSX_KeyShare_GenEccKey(ssl, ecc_kse);
+        if (ret != 0) {
+            /* No message, TLSX_KeyShare_GenEccKey() will do it. */
+            return ret;
```
Comment (dgarske) [inline src/tls.c:7267]: Will this leak `ecc_kse`?

File: src/tls.c [hunk 6]
```diff
@@ -8159,50 +8347,117 @@ static int TLSX_KeyShare_New(KeyShareEntry** list, int group, void *heap,
...
-                    "was enabled in liboqs.");
+    ret = wc_ecc_init_ex(&eccpubkey, ssl->heap, ssl->devId);
+    if (ret != 0) {
+        WOLFSSL_MSG("Could not do ECC public key initialization.");
         return MEMORY_E;
     }
 
-    if (len != kem->length_public_key) {
-        OQS_KEM_free(kem);
+    ecc_kse = XMALLOC(sizeof(*ecc_kse), ssl->heap, DYNAMIC_TYPE_TLSX);
```
Comment (dgarske) [inline src/tls.c:8379]: Must cast this `(KeyShareEntry*)`

**Review by dgarske — CHANGES_REQUESTED** [review]

File: src/tls.c (hunk 2 above)
Comment (anhu) [reply src/tls.c:7193]: @dgarske , what does "explicit start brace" mean?

Comment (dgarske) [reply src/tls.c:7193]:
```
static void findEccOqs(int *ecc, int *oqs, int group)
{
```

File: src/tls.c (hunk 6 above)
Comment (anhu) [reply src/tls.c:8379]: done. See next commit.

File: src/tls.c (hunk 4 above)
Comment (anhu) [reply src/tls.c:7252]: done. See next commit.

File: src/tls.c (hunk 1 above)
Comment (anhu) [reply src/tls.c:7161]: done. Please see next commit.

File: src/tls.c (hunk 2 above)
Comment (anhu) [reply src/tls.c:7193]: If it means drop the brace, then done. Please see next commit.

File: src/tls.c (hunk 3 above)
Comment (anhu) [reply src/tls.c:7235]: done...here and everywhere else.  Please see next commit.

File: src/tls.c (hunk 5 above)
Comment (anhu) [reply src/tls.c:7267]: Yes, it will.  Fix is done. Please see next commit.

**Review by dgarske — APPROVED** [review]

File: examples/server/server.c [hunk 7]
```diff
@@ -683,6 +683,58 @@ static void SetKeyShare(WOLFSSL* ssl, int onlyKeyShare, int useX25519,
                                 XSTRLEN("KYBER90S1024")) == 0) {
                 groups[count] = WOLFSSL_KYBER90S1024;
             }
+            else if (XSTRNCMP(oqsAlg, "P256_NTRUHPS2048509",
```
Comment (ejohnstown) [inline examples/server/server.c:686]: How does using strncmp, and strlen on a constant string help protect anything here?

             else if (XSTRNCMP(oqsAlg, “P521_NTRUHPS4096821”,
                                 XSTRLEN(“P521_NTRUHPS4096821")) == 0) {
                 groups[count] = WOLFSSL_P521_NTRU_HPS4096821;
             }

strncmp, I thought, used the length to be the length of the not-constant part, the buffer that you might read past the bounds on.

**Review by ejohnstown — CHANGES_REQUESTED** [review]

File: examples/server/server.c (hunk 7 above)
Comment (ejohnstown) [reply examples/server/server.c:686]: Given it is following the existing pattern, this is being overly picky. Our client and server examples should probably do a better job checking that the command line arguments are within the ballpark of the correct length and then not have to worry about this.

**Review by ejohnstown — APPROVED** [review]

File: src/tls.c [hunk 8]
```diff
@@ -8211,20 +8465,24 @@ static int server_generate_oqs_ciphertext(WOLFSSL* ssl,
...
-        ciphertext = NULL;
-
         keyShareEntry->ke = sharedSecret;
-        keyShareEntry->keLen = (word32) kem->length_shared_secret;
+        keyShareEntry->keLen = outlen + (word32)kem->length_shared_secret;
         sharedSecret = NULL;
+
+        XMEMCPY(ciphertext, ecc_kse->pubKey, ecc_kse->pubKeyLen);
+        keyShareEntry->pubKey = ciphertext;
+        keyShareEntry->pubKeyLen = (word32) ecc_kse->pubKeyLen +
```
Comment (ejohnstown) [inline src/tls.c:8474]: Clang on the Mac is complaining about loss of precision with this. `kem->length_ciphertext` is still 64-bit.

**Review by ejohnstown — CHANGES_REQUESTED** [review]

**Review by ejohnstown — APPROVED** [review]

**Comment by anhu** [comment]
@ejohnstown ....Thank you much!!

**Pull request merged**